### ISSUE TITLE
Node wishlist

### ISSUE THREAD

**Issue by ly29** [Sverchok/Sverchok#43]
Opening the gates for nodes that needs to be implemented, priority will be to the ones forcing new issues to be resolved. 

Also of course some nodes might needed to be pushed back until issues are resolved, either by working differently or by resolving, potentially systematic issues.

Please add things here.

Basic generator nodes
- [x] Torus
- [x] Plane/grid
- [ ] Cube
- [ ] Sphere
- [x] Cylinder

Math
- [ ] Formula node

List
- [x] List Repeat / Tile

Topology
- [ ] a simple `flip polygons` node (non bm) / with mask
    - using a robust ngon normal calculator that handles irregular shapes well (like [bmesh's normal calc](https://github.com/dfelinto/blender/blob/e6dda514bae1aa53724ac59afc5c98177bbdbdba/source/blender/bmesh/intern/bmesh_mesh.c#L468))
       or doing a 2d convex hull and geting the normalfrom that convex shape, 

The old Sverchok nodes
https://github.com/nortikin/sverchok/blob/master/index.md

New sverchok 0.5.9 nodes that should be remade for Rx.
- [ ] vector rewire
- [ ] exec node  (not high priority)

**Comment by ly29** [comment]
Also almost all nodes **should** start out as a script.

**Comment by nortikin** [comment]
that's the job for me, as i see.
Where is docs?

**Comment by nortikin** [comment]
ok, see
https://github.com/Sverchok/Sverchok/wiki/Node-Development-guide

**Comment by ly29** [comment]
Also it is just not an issue of porting all old nodes, sometimes they could very well be combined into one node.

**Comment by zeffii** [comment]
and numpy-ified, wrapping pydata for the basic primitives is not a great idea.

**Comment by ly29** [comment]
Yeah, for the boot strap stage is was ok, but now we should step things up a bit.

**Comment by nortikin** [comment]
will start from scripted node, from basics

**Comment by zeffii** [comment]
i'll do a numpy torus :)

**Comment by ly29** [comment]
Playing with the scripted node is a good starting point.

**Comment by zeffii** [comment]
oh. I miss the socket label showing the nested count.

**Comment by ly29** [comment]
I will see what I can do.

**Comment by ly29** [comment]
![depth](https://cloud.githubusercontent.com/assets/6241382/22862785/92db891c-f135-11e6-924c-4eac0928c18c.PNG)
@zeffii 
This shows the data depth, which isn't really the same but due in structure difference what is easily available.

**Comment by ly29** [comment]
A data structure viewer like stethoscope node would be good to have. Actually essential.

**Comment by ly29** [comment]
![object_count](https://cloud.githubusercontent.com/assets/6241382/22862866/4d2e8282-f137-11e6-9aa3-43eef531e234.PNG)
Total obj count without structure

**Comment by zeffii** [comment]
the last makes more sense . agreed about a viewer.

I'd like to propose an idea for a viewer / stethoscope.  https://github.com/Sverchok/Sverchok/issues/44

**Comment by zeffii** [comment]
this isn't quite possible yet.
![image](https://cloud.githubusercontent.com/assets/619340/22863324/4f4b2fa0-f13e-11e6-95a9-8cda7dee222a.png)

**Comment by zeffii** [comment]
oh.. i stand corrected. nice
![image](https://cloud.githubusercontent.com/assets/619340/22863355/f2132e5e-f13e-11e6-9947-c91822570721.png)

**Comment by ly29** [comment]
It needs a torus mode as well

**Comment by zeffii** [comment]
except for the cyclic flag / yes

**Comment by ly29** [comment]
It is a bit cobbled together however, just mostly wanted to try a node that would lower the complexity level. but also nice to have.

**Comment by zeffii** [comment]
- a simple `flip polygons` node (non bm)

**Comment by zeffii** [comment]
np.repeat / tile 
```python
import numpy as np
from svrx.nodes.node_base import node_func
from svrx.nodes.classes import NodeMathBase
from svrx.typing import Number, Int


@node_func(bl_idname=local_name, multi_label="Repeat", id=0, cls_bases=(NodeMathBase,))
def np_repeat(Values: Number = 0.0, Repeat: Int = 2) -> Number:
    return np.repeat(Values, Repeat)


@node_func(id=10)
def np_tile(Values: Number = 0.0, Repeat: Int = 2) -> Number:
    return np.tile(Values, Repeat)

```

**Comment by ly29** [comment]
Yeah repeat and tile surprised me as well when I looked into them. Nice with np cookbook, was thinking to do that

**Comment by zeffii** [comment]
```
 ------------------------------
 |  previous line-4        [use]| 
 |  previous line-3        [use]| 
 |  previous line-2        [use]| 
 |  previous line-1        [use]|
 |  previous line          [use]|
 |_>>> |__some command________|_|

```
with a button on the draw_buttons_ext to show a list of commands currentl implemented

**Comment by zeffii** [comment]
i'd been thinking about this for N1 but also it makes sense for svrx, a Console Node.
```
 ------------------------------
 |  previous line-4        [use]| 
 |  previous line-3        [use]| 
 |  previous line-2        [use]| 
 |  previous line-1        [use]|
 |  previous line          [use]|
 |_>>> |__some command________|_|

```
with a button on the draw_buttons_ext to show a list of commands currently implemented

**Comment by ly29** [comment]
That would be cool.

**Comment by zeffii** [comment]
List join A + B  ( + N....), for vertices, with option to 
- attach each array to the end 
- interleave A<sub>0</sub>, B<sub>0</sub>, A<sub>1</sub>, B<sub>1</sub>, ... A<sub>n</sub>, B<sub>n</sub>

**Comment by ly29** [comment]
@zeffii 
Will do today, corrected the signature for tile and repeat and removed the multiply constant in trig.

**Comment by zeffii** [comment]
> removed the multiply constant in trig.  

as long as there's still a convenient way to have multiples of pi/tau from one node.

**Comment by ly29** [comment]
I just removed the wrapper, it wasn't needed. It was only meant for when the constant was only the constant.
To allow code like

```
@node_func(...)
@constant
def e() -> Number:
    return np.e
```

So tau it becomes like this.
```python
@node_func(bl_idname=local_name, id=61)
def tau(n: Number = 1.0) -> Number:
    return np.pi * n * 2
```

**Comment by ly29** [comment]
The feeling you get when you realize that you have screwed up your git totally and the latest changes are out of sync... Not happy with the desktop client now, it should have at least warning for overwriting a repo with another one...

Only one hour of work luckily.

**Comment by zeffii** [comment]
oh man.. use `git for windows` has a bash-like terminal (ignore the UI part of the program)

**Comment by zeffii** [comment]
but losing work...  painful.

**Comment by ly29** [comment]
I could still access the changes in my text editor, so copied it out, nuked the git dir and started over.

**Comment by ly29** [comment]
![mix](https://cloud.githubusercontent.com/assets/6241382/23125645/1fcc03cc-f773-11e6-89eb-6ac33f4ab098.PNG)
I guess #55 is easier now.
I am holding off the N version but A + B join is here.

**Comment by zeffii** [comment]
great!  A+B is cool too  +N isn't so vital.

**Comment by ly29** [comment]
Okay so up now, called merge for now, had to handle uneven case in a resonable manner

**Comment by zeffii** [comment]
i'll make a SN with 3 if needed.

**Comment by zeffii** [comment]
hehe..or a `modes` version.. modes = 2,3,4,n

**Comment by zeffii** [comment]
could topology node have a mode where Line takes a skip rate, n , skep every n'th edge.
```
|                                   verts -|
|                                   edges -|
|          mode:  line skip n              |
|- verts                                   | 
|- n                                       |

```

**Comment by zeffii** [comment]
(ofcourse it _could_ - would it be suitable? )

**Comment by ly29** [comment]
Do it, worst case we can stick into a node  - strange topology generators

**Comment by zeffii** [comment]
parking here because i had no more time today
```python
from svrx.util.function import generator
from svrx.typing import IntP, Int, Bool

@node_script
@generator
def polyline_drop_n(
    verts: [Vertices] = Required, num: Int = 2, wrap: Bool = True) -> ([Vertices], [Edges], [Faces]):

    vertices = np.concatenate(verts)
    count = len(vertices)
    e_start = np.arange(0, count-1)
    e_stop = e_start + 1
    edges = np.array((e_start, e_stop)).T
    if wrap:
          edges = np.vstack([edges, [count-1, 0]])
    if num > 0:
        skip_list = list(range(num, count, num+1))
        edges = np.delete(edges, skip_list, axis=0)
    
    return vertices, edges, None
```

**Comment by zeffii** [comment]
Bool displays as an Int Property

**Comment by zeffii** [comment]
one thing that's bothered me about how Masking works.. is it needs lists of equal length as the input data,  i would like to be able to pass a list of `True` indices instead / or as option.

```
- data = [v0, v1, v2, v3, v4, v5, v6]
- mask = [0,3,6]

output data 
[v0, v3, v6] = True
[v1,v2,v4,v5] = False
```

**Comment by ly29** [comment]
Masking is worthy of an issue of its own. Very quick implementation...

**Comment by nortikin** [comment]
are we ignore level's workflow in SvRx?
How to deal with levels in general? I mean zip/join/shift and other list operations.

**Comment by ly29** [comment]
I handles levels internally but there aren't any node working with it. Yet. It is needed

**Comment by ly29** [comment]
Levels are a big part of the reason for Redux.

**Comment by nortikin** [comment]
Ok, thanks.
And sorry for my low activity, but my personal life became better (if one can compare), it takes lot of time.
